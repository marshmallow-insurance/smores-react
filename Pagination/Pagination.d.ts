import { FC } from 'react';
import { MarginProps } from '../utils/space';
export type PaginationProps = {
    /** Total number of records  */
    total: number;
    /** Number of records per page  */
    partition: number;
    /** Handle page change */
    handlePageChange: (page: number) => void;
    /** Sets current page number */
    currentPage?: number;
} & MarginProps;
export declare const Pagination: FC<PaginationProps>;
