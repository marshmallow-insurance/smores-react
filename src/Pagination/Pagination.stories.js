import React from 'react'
import { Pagination } from './Pagination'

export default {
  title: 'Pagination',
  component: Pagination,
  argTypes: { handlePageChange: { action: 'pageChanged' } },
}

const Template = (args) => <Pagination {...args} />

export const Default = Template.bind({})

Default.args = {
  total: 7,
  partition: 1,
}

export const LongerPagination = Template.bind({})

LongerPagination.args = {
  total: 11,
  partition: 1,
}
