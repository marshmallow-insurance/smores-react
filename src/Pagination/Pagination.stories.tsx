import React from 'react'
import { Pagination, PaginationProps } from './Pagination'

export default {
  title: 'Pagination',
  component: Pagination,
  argTypes: { handlePageChange: { action: 'pageChanged' } },
}

const Template = (props: PaginationProps) => <Pagination {...props} />

export const Default = Template.bind({})

Default.args = {
  total: 7,
  partition: 1,
}

export const LongerPagination = Template.bind({})

LongerPagination.args = {
  total: 11,
  partition: 1,
}
