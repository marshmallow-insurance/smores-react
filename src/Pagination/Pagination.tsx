import React, { FC, useEffect, useState } from 'react'
import styled from 'styled-components'

import { Box } from '../Box'
import { Text } from '../Text'
import { Icon } from '../Icon'

import { theme } from '../theme'
import { MarginProps } from '../utils/space'

const MAX_PAGES = 7
const MAX_ADDITIONAL_PAGES = Math.ceil(MAX_PAGES / 2)

export type PaginationProps = {
  /** Total number of records  */
  total: number
  /** Number of records per page  */
  partition: number
  /** Handle page change */
  handlePageChange: (page: number) => void
  /** Sets current page number */
  currentPage?: number
} & MarginProps

export const Pagination: FC<PaginationProps> = ({
  total,
  partition,
  handlePageChange,
  currentPage,
  ...marginProps
}) => {
  const [lastPage, setLastPage] = useState(0)
  const [activePage, setActivePage] = useState(1)
  const [pages, setPages] = useState<number[]>([])
  const [shownPages, setShownPages] = useState<number[]>([])
  const [showFirstDots, setShowFirstDots] = useState(false)
  const [showLastDots, setShowLastDots] = useState(false)

  useEffect(() => {
    // This is a rather hacky fix, in theory it should be listening to total, but because the total comes from the same endpoint as the data, it forces a reset to 1 every time
    // Using lastPage seems to bypass that
    if (currentPage) {
      setActivePage(currentPage)
    } else {
      setActivePage(1)
    }
  }, [lastPage, currentPage])

  useEffect(() => {
    const numberOfPages = Math.ceil(total / partition)
    if (numberOfPages > 1) {
      setLastPage(numberOfPages)
    } else if (numberOfPages === 1) {
      // Sets last page to be null to avoid the last page being kept from previous state.
      setLastPage(0)
    }
    if (numberOfPages > 2) {
      const allPages = Array.from({ length: numberOfPages }, (_, i) => i + 1)
      setPages(allPages.slice(1, numberOfPages - 1))
    } else {
      // Sets pages to empty to avoid using the previous state if pages set before.
      setPages([])
    }
  }, [total, partition])

  useEffect(() => {
    if (lastPage <= MAX_PAGES) {
      setShownPages(pages)
    } else if (activePage <= MAX_ADDITIONAL_PAGES) {
      setShownPages(pages.slice(0, MAX_ADDITIONAL_PAGES))
    } else if (activePage > lastPage - MAX_ADDITIONAL_PAGES) {
      setShownPages(pages.slice(pages.length - MAX_ADDITIONAL_PAGES))
    } else {
      const activePageIndex = pages.indexOf(activePage)
      setShownPages(pages.slice(activePageIndex - 1, activePageIndex + 2))
    }
  }, [lastPage, activePage, pages])

  useEffect(() => {
    setShowFirstDots(shownPages[0] !== pages[0])
    setShowLastDots(
      shownPages[shownPages.length - 1] !== pages[pages.length - 1],
    )
  }, [shownPages, pages])

  const movePage = (page: number) => {
    if (page > 0 && page <= lastPage) {
      setActivePage(page)
      handlePageChange(page)
    }
  }

  return (
    <Container flex direction="row" {...marginProps}>
      {activePage - 1 > 0 && (
        <PageBox onClick={() => movePage(activePage - 1)}>
          <Icon size={24} render="arrow" />
        </PageBox>
      )}
      <PageBox active={activePage === 1} onClick={() => movePage(1)}>
        <Text typo={activePage === 1 ? 'desc-medium' : 'desc-light'} tag="p">
          1
        </Text>
      </PageBox>
      {showFirstDots && (
        <Box flex justifyContent="center">
          <Text tag="p">...</Text>
        </Box>
      )}
      {shownPages.map((i) => {
        return (
          <PageBox
            active={activePage === i}
            key={i}
            onClick={() => movePage(i)}
          >
            <Text
              typo={activePage === i ? 'desc-medium' : 'desc-light'}
              tag="p"
            >
              {i}
            </Text>
          </PageBox>
        )
      })}
      {showLastDots && (
        <Box flex justifyContent="center">
          <Text tag="p">...</Text>
        </Box>
      )}
      {Boolean(lastPage) && (
        <PageBox
          active={activePage === lastPage}
          onClick={() => movePage(lastPage)}
        >
          <Text
            typo={activePage === lastPage ? 'desc-medium' : 'desc-light'}
            tag="p"
          >
            {lastPage}
          </Text>
        </PageBox>
      )}
      {activePage + 1 <= lastPage && (
        <PageBox onClick={() => movePage(activePage + 1)}>
          <Icon size={18} render="arrow" rotate={180} />
        </PageBox>
      )}
    </Container>
  )
}

interface IPageBox {
  active?: boolean
}

const Container = styled(Box)`
  align-items: center;
  height: 24px;
  > div {
    width: 24px;
    margin-right: 4px;

    &:last-child {
      margin-right: 0;
    }
  }
`

const PageBox = styled(Box)<IPageBox>`
  background: ${({ active }) => (active ? theme.colors.coconut : 'none')};
  cursor: pointer;
  display: flex;
  justify-content: center;
  border-radius: 50%;
`
