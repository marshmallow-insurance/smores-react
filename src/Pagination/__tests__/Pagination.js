import React from 'react'
import { render } from '@testing-library/react'
import 'jest-styled-components'

import { Pagination } from '../Pagination'

test('renders', () => {
  const { container } = render(
    <Pagination total={11} partition={1} handlePageChange={() => {}} />,
  )
  expect(container.firstChild).toMatchSnapshot()
})
