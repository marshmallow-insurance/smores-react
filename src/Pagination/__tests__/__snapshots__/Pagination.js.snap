// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 18px;
  height: 18px;
  -webkit-transform: rotate(180deg);
  -ms-transform: rotate(180deg);
  transform: rotate(180deg);
}

.c3 {
  font-size: 14px;
  line-height: 20px;
  font-weight: 500;
  text-align: left;
  cursor: inherit;
  color: #292924;
}

.c5 {
  font-size: 14px;
  line-height: 20px;
  font-weight: 400;
  text-align: left;
  cursor: inherit;
  color: #292924;
}

.c6 {
  font-size: 16px;
  font-weight: 400;
  line-height: 20px;
  text-align: left;
  cursor: inherit;
  color: #292924;
}

.c1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 > button {
  margin-right: 4px;
}

.c1 > button:last-child {
  margin-right: 0;
}

.c2 {
  background: #F1E9DC;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border-radius: 50%;
  width: 24px;
  height: 24px;
}

.c2:hover {
  background-color: #F8F2EA;
}

.c2:active {
  background-color: #F1E9DC;
}

.c2:focus-visible {
  outline: 2px solid #292924;
  outline-offset: 2px;
}

.c4 {
  background: #FFFFFF;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border-radius: 50%;
  width: 24px;
  height: 24px;
}

.c4:hover {
  background-color: #F8F2EA;
}

.c4:active {
  background-color: #F1E9DC;
}

.c4:focus-visible {
  outline: 2px solid #292924;
  outline-offset: 2px;
}

.c7 {
  cursor: pointer;
}

.c7:focus-visible {
  outline: 2px solid #292924;
  outline-offset: 2px;
}

@media (min-width:0px) {
  .c0 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
  }
}

@media (min-width:768px) {
  .c3 {
    font-size: 16px;
    line-height: 23px;
  }
}

@media (min-width:768px) {
  .c5 {
    font-size: 16px;
    line-height: 23px;
  }
}

<div
  class="c0 c1"
  direction="row"
>
  <button
    class="c2"
  >
    <p
      class="c3"
      color="liquorice"
      cursor="inherit"
      title=""
    >
      1
    </p>
  </button>
  <button
    class="c4"
  >
    <p
      class="c5"
      color="liquorice"
      cursor="inherit"
      title=""
    >
      2
    </p>
  </button>
  <button
    class="c4"
  >
    <p
      class="c5"
      color="liquorice"
      cursor="inherit"
      title=""
    >
      3
    </p>
  </button>
  <button
    class="c4"
  >
    <p
      class="c5"
      color="liquorice"
      cursor="inherit"
      title=""
    >
      4
    </p>
  </button>
  <button
    class="c4"
  >
    <p
      class="c5"
      color="liquorice"
      cursor="inherit"
      title=""
    >
      5
    </p>
  </button>
  <button
    class="c4"
  >
    <p
      class="c6"
      color="liquorice"
      cursor="inherit"
      title=""
    >
      ...
    </p>
  </button>
  <button
    class="c4"
  >
    <p
      class="c5"
      color="liquorice"
      cursor="inherit"
      title=""
    >
      11
    </p>
  </button>
  <button
    class="c7"
  >
    <span
      class="c8"
      color="liquorice"
      rotate="180"
      size="18"
    >
      <svg
        fill="none"
        height="100%"
        viewBox="0 0 24 24"
        width="100%"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M2.087 12.401c.014.033.042.058.06.089a.99.99 0 00.177.238c.008.007.01.018.018.025l8 7a1 1 0 101.316-1.505L5.661 13H21a1 1 0 100-2H5.661l5.997-5.248a1 1 0 10-1.316-1.504l-8 7c-.008.007-.011.018-.018.025a.99.99 0 00-.177.238c-.018.031-.045.056-.06.089a.97.97 0 000 .801z"
          fill="#292924"
        />
      </svg>
    </span>
  </button>
</div>
`;
